File: src/cpp/src/padding.cpp
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

#include "array_view.hpp"
#include "padding.hpp"

#include "ascii_char_set.hpp"
#include "ast.hpp"

#include <algorithm>
#include <array>
#include <cstddef>
#include <cstdint>
#include <memory>
#include <utility>
#include <vector>

namespace logmu::detail::padding
{
	namespace
	{
		constexpr std::array<std::uint64_t, 16> powers_of_ten = []
		{
			std::array<std::uint64_t, 16> powers{};
			powers[0] = 1;

			for (std::size_t i = 1; i < powers.size(); ++i)
			{
				powers[i] = powers[i - 1] * 10;
			}

			return powers;
		}();

		ast_ptr zero(std::size_t offset)
		{
			return std::make_shared<ast_chars>(ascii_char_set::single_character('0'), offset);
		}

		/// One padding position: a mandatory zero, or the unified element its mark stands for.
		ast_ptr padding_position(char mark, std::size_t offset)
		{
			if (mark == '\0')
			{
				return zero(offset);
			}

			// The expansions are the ones the specification gives: 0! is 0!!, which is 0?!(0),
			// and 0? is 0!?, which is 0?!().
			ast_ptr subject = std::make_shared<ast_optional>(zero(offset), offset);
			const bool reproduced = mark == '!';

			ast_ptr rendering = reproduced
				? zero(offset)
				: ast_ptr(std::make_shared<ast_empty>(offset));

			const unified_form form = reproduced ? unified_form::reproduced : unified_form::dropped;

			return std::make_shared<ast_unified>(std::move(subject), std::move(rendering), form, offset);
		}

		/// The alternative for values of one width: its padding, then the values themselves.
		ast_ptr alternative(std::uint64_t first, std::uint64_t last, int width, int low_digit_count, array_view<char> marks, std::size_t offset)
		{
			const int pad_count = std::max(0, low_digit_count - width);
			ast_ptr values = std::make_shared<ast_decimal_range>(first, width, last, width, offset);

			if (pad_count == 0)
			{
				return values;
			}

			std::vector<ast_ptr> parts;
			parts.reserve(static_cast<std::size_t>(pad_count) + 1);

			for (int position = 0; position < pad_count; ++position)
			{
				parts.push_back(padding_position(marks[static_cast<std::size_t>(position)], offset));
			}

			parts.push_back(std::move(values));

			return std::make_shared<ast_sequence>(std::move(parts), offset);
		}
	}

	ast_ptr expand(
		std::uint64_t low,
		int low_digit_count,
		std::uint64_t high,
		int high_digit_count,
		array_view<char> marks,
		std::size_t offset)
	{
		std::vector<ast_ptr> alternatives;
		alternatives.reserve(static_cast<std::size_t>(high_digit_count));

		for (int width = 1; width <= high_digit_count; ++width)
		{
			const std::uint64_t lowest = width == 1 ? 0 : powers_of_ten[static_cast<std::size_t>(width - 1)];
			const std::uint64_t highest = powers_of_ten[static_cast<std::size_t>(width)] - 1;

			const std::uint64_t first = std::max(low, lowest);
			const std::uint64_t last = std::min(high, highest);

			if (first > last)
			{
				continue;
			}

			alternatives.push_back(alternative(first, last, width, low_digit_count, marks, offset));
		}

		return alternatives.size() == 1
			? alternatives[0]
			: std::make_shared<ast_alternation>(std::move(alternatives), offset);
	}
}

File: src/js/lib/padding.js
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

import { AsciiCharSet } from './ascii-char-set.js';
import {
	AstAlternation,
	AstChars,
	AstDecimalRange,
	AstEmpty,
	AstOptional,
	AstSequence,
	AstUnified,
	UnifiedForm,
} from './ast.js';

/**
 * The expansion of a decimal range whose lower bound carries marks, as in `#[0!0!0-105]`.
 *
 * A mark is shorthand and nothing downstream of the parser sees one. `0!` becomes `0!!` and `0?`
 * becomes `0!?`, so the language, the well-formedness rules and the encoding all apply to the
 * expansion and none of them needs a case for padding. W2 falls out of this in the same way it
 * does for a case fold: a marked range inside the operand of a `!` puts a `!` there, which the
 * existing nesting check refuses.
 *
 * The shape is one alternative per width the value itself takes, with the padding positions that
 * apply to that width written out in front. A padding position is mandatory where it carries no
 * mark, which is what makes an unmarked leading zero go on setting a minimum width.
 *
 * A unified element contributes its rendering whether or not its subject matched anything, so a
 * run of padding contributes a fixed string. That is what keeps `07` to one canonical form under
 * `0!! 0!!`, where either of the two could have been the one that matched.
 *
 * Every node made here takes the pattern offset of the range it replaces, so a fault found in an
 * expansion still points at what the author wrote.
 */

/** Ten to the power of the index, to the fifteen digits a bound may have. */
const POWERS_OF_TEN = Object.freeze(
	Array.from({ length: 16 }, (_unused, exponent) => 10 ** exponent));

/**
 * Expands a marked decimal range into an alternation of ordinary elements.
 *
 * @param {number} low The value of the lower bound.
 * @param {number} lowDigitCount The digits the lower bound was written with.
 * @param {number} high The value of the upper bound.
 * @param {number} highDigitCount The digits the upper bound was written with.
 * @param {string[]} marks One entry per digit of the lower bound, each `'!'`, `'?'` or the empty
 *   string where that digit carries no mark. The array may be longer than the bound.
 * @param {number} offset The offset of the range in the pattern, for diagnostics.
 * @returns {import('./ast.js').Ast} The expansion.
 */
export function expandPadding(low, lowDigitCount, high, highDigitCount, marks, offset) {
	const alternatives = [];

	for (let width = 1; width <= highDigitCount; ++width) {
		const first = Math.max(low, width === 1 ? 0 : POWERS_OF_TEN[width - 1]);
		const last = Math.min(high, POWERS_OF_TEN[width] - 1);

		if (first > last) { continue; }

		alternatives.push(alternative(first, last, width, lowDigitCount, marks, offset));
	}

	return alternatives.length === 1
		? alternatives[0]
		: at(new AstAlternation(alternatives), offset);
}

/**
 * The alternative for values of one width: its padding, then the values themselves.
 *
 * @param {number} first The lowest value of this width.
 * @param {number} last The highest.
 * @param {number} width The digits the value itself takes.
 * @param {number} lowDigitCount The digits the lower bound was written with.
 * @param {string[]} marks The marks, one per digit of the lower bound.
 * @param {number} offset The offset of the range in the pattern.
 * @returns {import('./ast.js').Ast} The alternative.
 */
function alternative(first, last, width, lowDigitCount, marks, offset) {
	const padCount = Math.max(0, lowDigitCount - width);
	const values = at(new AstDecimalRange(first, width, last, width), offset);

	if (padCount === 0) { return values; }

	const parts = [];

	for (let position = 0; position < padCount; ++position) {
		parts.push(paddingPosition(marks[position], offset));
	}

	parts.push(values);

	return at(new AstSequence(parts), offset);
}

/**
 * One padding position: a mandatory zero, or the unified element its mark stands for.
 *
 * @param {string} mark The mark, `'!'`, `'?'` or the empty string.
 * @param {number} offset The offset of the range in the pattern.
 * @returns {import('./ast.js').Ast} The element.
 */
function paddingPosition(mark, offset) {
	if (!mark) { return zero(offset); }

	// The expansions are the ones the specification gives: 0! is 0!!, which is 0?!(0), and 0? is
	// 0!?, which is 0?!().
	const reproduced = mark === '!';
	const subject = at(new AstOptional(zero(offset)), offset);
	const rendering = reproduced ? zero(offset) : at(new AstEmpty(), offset);
	const form = reproduced ? UnifiedForm.Reproduced : UnifiedForm.Dropped;

	return at(new AstUnified(subject, rendering, form), offset);
}

/**
 * A literal zero.
 *
 * @param {number} offset The offset of the range in the pattern.
 * @returns {import('./ast.js').Ast} The element.
 */
function zero(offset) {
	return at(new AstChars(AsciiCharSet.fromSingleChar(0x30)), offset);
}

/**
 * Puts a pattern offset on a node, which is what makes a fault in an expansion point at what the
 * author wrote.
 *
 * @template {import('./ast.js').Ast} T
 * @param {T} node The node.
 * @param {number} offset The offset.
 * @returns {T} The node.
 */
function at(node, offset) {
	node.patternOffset = offset;

	return node;
}
